ipatests/test_integration/test_acme.py

```diff
@@ -0,0 +1,236 @@
+        # ACME service should return 503 because it was not enabled yet.
+        self.clients[0].run_command(
+            ['curl', '--fail', self.acme_server],
+            ok_returncode=22,
+        )
+
+    def test_enable_acme_service(self):
+        self.master.run_command(['ipa-acme-manage', 'enable'])
+
+        # wait a short time for Dogtag ACME service to observe config
+        # change and reconfigure itself to service requests
+        time.sleep(3)
+
+        # should succeed now
+        self.clients[0].run_command(['curl', '--fail', self.acme_server])
+
+    ###############
+    # Certbot tests
+    ###############
+
+    @pytest.mark.skipif(skip_certbot_tests, reason='certbot not available')
+    def test_certbot_register(self):
+        # service is enabled; registration should succeed
+        self.clients[0].run_command(
+            [
+                'certbot',
+                '--server', self.acme_server,
+                'register',
+                '-m', 'nobody@example.com',
+                '--agree-tos',
+                '--no-eff-email',
+            ],
+        )
+
+    @pytest.mark.skipif(skip_certbot_tests, reason='certbot not available')
+    def test_certbot_certonly_standalone(self):
+        # Get a cert from ACME service using HTTP challenge and Certbot's
+        # standalone HTTP server mode
+        self.clients[0].run_command(
+            [
+                'certbot',
+                '--server', self.acme_server,
+                'certonly',
+                '--domain', self.clients[0].hostname,
+                '--standalone',
+            ],
+        )
+
+    @pytest.mark.skipif(skip_certbot_tests, reason='certbot not available')
+    def test_certbot_revoke(self):
+        # Assume previous certonly operation succeeded.
+        # Read certificate to learn serial number.
+        cert_path = \
+            f'/etc/letsencrypt/live/{self.clients[0].hostname}/cert.pem'
+        data = self.clients[0].get_file_contents(cert_path)
+        cert = x509.load_pem_x509_certificate(data, backend=default_backend())
+
+        # revoke cert via ACME
+        self.clients[0].run_command(
+            [
+                'certbot',
+                '--server', self.acme_server,
+                'revoke',
+                '--cert-name', self.clients[0].hostname,
+                '--delete-after-revoke',
+            ],
+        )
+
+        # check cert is revoked (kinit already performed)
+        result = self.master.run_command(
+            ['ipa', 'cert-show', str(cert.serial_number), '--raw']
+        )
+        assert 'revocation_reason:' in result.stdout_text
+
+    @pytest.mark.skipif(skip_certbot_tests, reason='certbot not available')
+    def test_certbot_dns(self):
+        # Assume previous revoke operation succeeded and cert was deleted.
+        # We can now request a new certificate.
+
+        # Get a cert from ACME service using HTTP challenge and Certbot's
```

Comment:
nitpick: need to correct challenge as DNS.

Reply:
Will be fixed in next push.